### PR TITLE
Speed Up info_locale

### PR DESCRIPTION
**Pull request by Carterpersall**
- Made info_locale speedy
  - Takes the registry entries at HKCU:Control Panel\International to get region code and language code, and passes them into two very large hash tables to get the accompanying region and language values.
  - Maybe: Find a way to circumvent the very large table
  - 3251 ms -> 12.43 ms
    - 262x Faster

### CODE REVIEW

**Comment by Carterpersall** [comment]
Unless you or I think of anything else, I think this pull request is ready for review.

**Comment by rashil2000** [comment]
Could you arrange the columns (with appropriate spaces) so that they're readable (even when cluttered)? Right now it looks like a giant blob of text.

Also do that for the region table.

**Comment by Carterpersall** [comment]
> Could you arrange the columns (with appropriate spaces) so that they're readable (even when cluttered)? Right now it looks like a giant blob of text.
> 
> Also do that for the region table.

@rashil2000  Done

**Comment by rashil2000** [comment]
Can you post before and after numbers for time taken, on both PowerShell 5.1 and 7.2.6

**Comment by rashil2000** [comment]
On my machine:

If I remove the International module workaround (which I will do in the master branch shortly), then there's no performance difference between that and the implementation in this PR, for both PS 5.1 and 7.2.6

**Comment by Carterpersall** [comment]
Order is Old -> New -> Pull Request

Heres the files of the results I got: 
[Script Profiles.zip](https://github.com/kiedtl/winfetch/files/9568569/Script.Profiles.zip)

The files in the ZIP can be imported to https://www.speedscope.app/ which I use to profile scripts using https://github.com/nohwnd/Profiler

Powershell 5.1.2200.832
- 106.1654 ms -> 92.7306 ms -> 26.9758 ms

Powershell 7.2.6
- 2519.2612 ms -> 134.0106 ms -> 12.43 ms

**Comment by rashil2000** [comment]
Unfortunately, I'm still not able to reproduce any difference. I isolated the `info_locale` function into a file, and ran some benchmarks:

image

[Simple.ps1.txt](https://github.com/kiedtl/winfetch/files/9569780/Simple.ps1.txt)

[Lookup_table.ps1.txt](https://github.com/kiedtl/winfetch/files/9569782/Lookup_table.ps1.txt)

**Comment by Carterpersall** [comment]
Here's what I got on my laptop:
image

My laptop has a Ryzen 9 5900HX throttled down to 1.09 GHz which I do to preserve battery life and is the state in which all my PowerShell speed tests are run. I do this because I've found that when PowerShell runs scripts on slow systems, seemingly small performance decreases can sometimes be blown up to big ones.

I'm also not a big fan of using hyperfine to test the performance of individual functions, as the tests include the launch speed of the PowerShell instance itself, which can be volatile and cause a large range in the results, which can make small performance gains very hard to notice.

**Comment by rashil2000** [comment]
> as the tests include the launch speed of the PowerShell instance itself

I somewhat agree, but in this case, we use the `-w` parameter in hyperfine to warm-up the .NET runtime, so that shell launch times become mostly consistent. 

The reason I did not use PowerShell's built-in `Measure-Command` is that PowerShell caches the results of cmdlets (they only take time the first time, subsequent runs are instantaneous), and so it does not really give a good estimate of performance.

> which can make small performance gains very hard to notice.

In this PR the gains are pretty small to notice already (after the removal of International module, that is, which was the main bottleneck) - and I'm not sure the added complexity in the script (w.r.t. readability and length) is worth it.

**Comment by Carterpersall** [comment]
> I somewhat agree, but in this case, we use the `-w` parameter in hyperfine to warm-up the .NET runtime, so that shell launch times become mostly consistent.

Even after warmup, the launch speed of PowerShell can vary by 30-40 ms, which is small but can make a difference

> The reason I did not use PowerShell's built-in `Measure-Command` is that PowerShell caches the results of cmdlets (they only take time the first time, subsequent runs are instantaneous), and so it does not really give a good estimate of performance.
 
By running `pwsh -noprofile {Measure-Command {Insert stuff here}}` works as it spawns a fresh instance for Measure-Command, eliminating the issue. Also the Profiler module uses PowerShell's built-in debug tracer, making it accurate down to 10 microseconds.


> In this PR the gains are pretty small to notice already (after the removal of International module, that is, which was the main bottleneck) - and I'm not sure the added complexity in the script (w.r.t. readability and length) is worth it.

For me, according to the profiler I got 10x better speeds(134.0106 ms -> 12.43 ms), and according to hyperfine I got 33% better speeds, which I think is fairly significant. As for the readability impacts, while I do agree the size of the tables are annoying, I don't think readability is impacted too much as most people will read the comments and scroll past the tables.

**Comment by rashil2000** [comment]
I ran the files through the Profiler and got these results:
[Simple.speedscope.json.txt](https://github.com/kiedtl/winfetch/files/9570515/Simple.speedscope.json.txt)
[Lookup_table.speedscope.json.txt](https://github.com/kiedtl/winfetch/files/9570518/Lookup_table.speedscope.json.txt)


image

image


Honestly, I'm confused again

**Comment by rashil2000** [comment]
@jcwillox could you also run the files mentioned in https://github.com/kiedtl/winfetch/pull/154#issuecomment-1247272030 and post your results?

**Comment by Carterpersall** [comment]
Yeah, I've found that when PowerShell is running at a very fast speed, speed gains seem to become negligible. But when PowerShell runs on a slow processor, some things just start running significantly slower for whatever reason. But the commands that do run slower are consistently slower and can be optimized by replacing those commands with fast alternatives.

**Comment by rashil2000** [comment]
> But the commands that do run slower are consistently slower and can be optimized by replacing those commands with fast alternatives.

The problem here is that there isn't one command that always runs slower. Because of invariant factors (shell startup, cmdlet caching or whatever), I see that any of them can be slower (at least on my machine). I repeated the experiments by unplugging the power cable and enabling the aggressive power-saver mode (which severely restricts CPU performance), my results are the same - sometimes Simple.ps1 wins, sometimes Lookup_table.ps1 wins.

Which leads me to the conclusion that on an average, they're equally performant.

**Comment by Carterpersall** [comment]
Maybe the speed gains vary from computer to computer due to many factors from how slow the computer is to driver and BIOS overhead. But for me, certain commands are always much slower (specifically anything CIM or WMI) making this implementation almost always faster for me and significantly so.

File: winfetch.ps1 [hunk 1]
```diff
@@ -963,22 +963,402 @@ function info_battery {
...
+        "zh-Hans-MO" = "Chinese (Simplified Han, Macao SAR)";         "zh-Hant" = "Chinese (Traditional)";                               "zh-HK" = "Chinese (Traditional, Hong Kong S.A.R.)";                                                                                                                                            
+        "zh-MO" = "Chinese (Traditional, Macao S.A.R.)";              "zh-SG" = "Chinese (Simplified, Singapore)";                       "zh-TW" = "Chinese (Traditional, Taiwan)";                                                                                                                                            
+        "zu" = "isiZulu";                                             "zu-ZA" = "isiZulu (South Africa)"
+    }
+
+    # Get the current user's language and region using the registry
+    $Region = $localeLookup[(Get-ItemProperty -Path "HKCU:Control Panel\International\Geo").Nation]
+    # Iterate through registry key in case multiple languages are configured
+    (Get-ItemProperty -Path "HKCU:Control Panel\International\User Profile").Languages | ForEach-Object{
```
Comment (rashil2000) [inline winfetch.ps1:1355]:
```suggestion
    (Get-ItemProperty -Path "HKCU:Control Panel\International\User Profile").Languages | ForEach-Object {
```

**Comment by rashil2000** [comment]
Alright, let's merge it then

**Pull request merged**